### DISCUSSION TITLE
`it` doesn't work inside `forEach` loop

### DISCUSSION THREAD

**Discussion by 0618** [cypress-io/cypress#24345] [Questions and Help] [answered]
Can anyone help me understand why doesn't the test inside the `forEach` run? Thanks!

```ts
describe("all the links", () => {
    let allLinks = [];
    before(() => {
      cy.task('readSitemapLinks').then((links: string[]) => {
        allLinks = allLinks.concat(links);
      });
    });
    it("should have 100 links", () => {  // This test runs
      expect(allLinks.length).to.eq(100);
    })

    allLinks.forEach((link) => {
      it('the link should work', () => {  // These tests never run
        cy.request(link ? link : '/').then(({ status }) => {
          expect(status).to.eq(200);
        });
      });
  });
})
```

**Comment by 0618** [comment]
but `for` loop works

```ts
for (let i = 0; i < 3; i++) {
   describe("something", ()=> {
     it("some tests", () => {})
  })
}
```

**Comment by filiphric** [comment, marked as answer]
you are running into the asynchronicity issue. think of `before()` block as a part of your `it()` blocks. basically, the reason your tests aren’t running is because you are generating your tests using empty array. that array does not get filled until `before()` hook runs. but since `before()` hook is essentially a part of your `it()`, those tests will never get generated.

you can easily fix this issue by moving your `allLinks.forEach()` function inside your `it()` block

**Reply by 0618** [reply]
That makes sense. Because `for (let i = 0; i < allLinks.length; i++) {...` also doesn't work.

**Reply by Hexronimo** [reply]
I'm new to Cypress, so maybe I understand it wrongly, but I think there is difference between loop inside it and outside it:

- _it_ inside the loop will create separate test for each link
- _it_ outside the loop will create single test which will break on the first link with an assertion error and do not check the rest

I also encountered this problem and I don't understand how to solve it yet. I need to collect links and test them all without breaking on error.
